test/perf_counters_gtest.cc

```diff
@@ -85,110 +81,135 @@ TEST(PerfCountersTest, Read2Counters) {
   if (!PerfCounters::kSupported) {
     GTEST_SKIP() << "Test skipped because libpfm is not supported.\n";
   }
   EXPECT_TRUE(PerfCounters::Initialize());
-  std::vector<PerfCounters> counters;
+  std::vector<std::shared_ptr<PerfCounters>> counters;
   counters.reserve(6);
   for (int i = 0; i < 6; i++)
     counters.push_back(PerfCounters::Create({kGenericPerfEvent1}));
   PerfCounterValues values(1);
-  EXPECT_TRUE(counters[0].Snapshot(&values));
-  EXPECT_FALSE(counters[4].Snapshot(&values));
-  EXPECT_FALSE(counters[5].Snapshot(&values));
+  EXPECT_TRUE(counters[0]->Snapshot(&values));
+  EXPECT_TRUE(counters[4]->Snapshot(&values));
+  EXPECT_TRUE(counters[5]->Snapshot(&values));
 }
 
 TEST(PerfCountersTest, CreateExistingMeasurements) {
-  // The test works (i.e. causes read to fail) for the assumptions
-  // about hardware capabilities (i.e. small number (3-4) hardware
-  // counters) at this date,
-  // the same as previous test ReopenExistingCounters.
   if (!PerfCounters::kSupported) {
     GTEST_SKIP() << "Test skipped because libpfm is not supported.\n";
   }
   EXPECT_TRUE(PerfCounters::Initialize());
-  std::vector<PerfCountersMeasurement> perf_counter_measurements;
+
+  const int kMaxCounters = 1;
+  const std::vector<std::string> kMetrics{"cycles"};
+
+  std::vector<PerfCountersMeasurement> perf_counter_measurements(kMaxCounters,
+                                                                 kMetrics);
   std::vector<std::pair<std::string, double>> measurements;
 
-  perf_counter_measurements.reserve(10);
-  for (int i = 0; i < 10; i++)
-    perf_counter_measurements.emplace_back(
-        std::vector<std::string>{kGenericPerfEvent1});
+  // Start all together
+  int max_counters = kMaxCounters;
+  for (int i = 0; i < kMaxCounters; ++i) {
+    PerfCountersMeasurement& counter(perf_counter_measurements[i]);
+    EXPECT_TRUE(counter.IsValid());
+    EXPECT_EQ(counter.num_counters(), 1);
+    if (!counter.IsValid()) {
+      max_counters = i;
+      break;
+    }
+    counter.Start();
+  }
 
-  perf_counter_measurements[0].Start();
-  EXPECT_TRUE(perf_counter_measurements[0].Stop(measurements));
+  ASSERT_GE(max_counters, 1);
 
-  measurements.clear();
-  perf_counter_measurements[8].Start();
-  EXPECT_FALSE(perf_counter_measurements[8].Stop(measurements));
+  // Start all together
+  for (int i = 0; i < max_counters; ++i) {
+    PerfCountersMeasurement& counter(perf_counter_measurements[i]);
+    EXPECT_TRUE(counter.Stop(measurements));
+  }
 
-  measurements.clear();
-  perf_counter_measurements[9].Start();
-  EXPECT_FALSE(perf_counter_measurements[9].Stop(measurements));
+  // Start/stop individually
+  for (int i = 0; i < max_counters; ++i) {
+    PerfCountersMeasurement& counter(perf_counter_measurements[i]);
+    measurements.clear();
+    counter.Start();
+    EXPECT_TRUE(counter.Stop(measurements));
+  }
 }
 
+using benchmark::ChronoClockNow;
+
 size_t do_work() {
+  const double kMinimumElapsedSeconds = 0.25;
```

Comment:
i'm not a fan of explicit waits in tests as they tend to get flaky and surprising (and obviously slow).

the _right_ way to do this i think is to use `DoNotOptimize` on `res`.

Reply:
You got another case yesterday that I commented where the compiler was smart enough to collapse the loop even though the DoNotOptimize worked. That happened here too. The compiler is just outsmarting our assumptions. In this case, that was the reason for the tests failing in release mode.

Reply:
try `__attribute__((optnone))` on `doWork`. It's `__attribute__((optimize(0))` in gcc. The test shouldn't run on windows anyway, so the msvc toolchain alternative doesn't apply here.

Reply:
Are you guys okay defining BENCHMARK_DONT_OPTIMIZE in include/benchmark.h along with BENCHMARK_ALWAYS_INLINE and BENCHMARK_NOEXCEPT just to have the compiler-specific defines all in one place?

Reply:
Having trouble with MSVC though... it does not accept single-function attributes like gcc and clang so we would be forced to use two OFF/ON macros like [here](https://godbolt.org/z/e5Erhzf6Y). However that requires the standard `_Pragma()` form to be used which is somewhat new and might break old compilers.

Reply:
the test shouldn't even run on msvc, because of that `!PerfCounters::kSupported` test, so maybe on msvc don't add any attribute just to make the compiler happy?

Reply:
True. Ends up it seems that it uses the same syntax of the Intel compiler. Not surprised by that actually.

Reply:
it looks like we ended up using _Pragma instead of __attribute__ .. i don't know enough about this area but why is that? __attribute__ should be well-supported by the compilers we care about.

Reply:
I would file that in the bad try shelf. That was because MSVC (always you Microsoft!) does not have an equivalent attribute but it has a pragma, so my solution was to make a pair of ON/OFF macros. We could have both or return to the previous attribute macro with the caveat that it will not work on MSVC.

Reply:
Correction - MSVC does not have a do-not-optimize attribute, it has a line-level `#pragma` only.

Reply:
Committed a revert change. Please have a look. Thank you.